### PR TITLE
Separate liquidity flags

### LINKED ISSUE #957: Consider disabling unbalanced add/remove separately
It could be useful to disable only unbalanced removes, so that hooks could charge a fee on exit safely, yet still support unbalanced adds.

### PR DESCRIPTION
**Pull request by EndymionJkb**
# Description

By default, the Vault supports all non-custom add/remove liquidity types: single token, unbalanced, and proportional. However, there are use cases where unbalanced liquidity operations may be unsafe or unsupported.

For instance, pricing in pools that support donation could be manipulated if it were possible to add or remove liquidity unbalanced. Also, if the hook enables adjusting swap deltas, it cannot be registered to a pool that supports unbalanced liquidity, or the fees (always on the calculated amount), might have to be calculated in BPT, which isn't supported. (The Vault prevents this.)

However, this check might be too restrictive. Hooks with exit fees should enforce proportional withdrawals to ensure proper fee accounting: but might still allow unbalanced adds, as this is the default in the UI and would be most convenient. This PR splits the `disableUnbalancedLiquidity` flag in LiquidityManagement into two separate flags: `disableAddLiquidityUnbalanced` and `disableRemoveLiquidityUnbalanced`.

## Type of change

- [ ] Bug fix 
- [X] New feature 
- [ ] Breaking change 
- [ ] Dependency changes
- [ ] Code refactor / cleanup
- [ ] Optimization: [ ] gas / [ ] bytecode
- [ ] Documentation or wording changes
- [ ] Other

## Checklist:

- [X] The diff is legible and has no extraneous changes
- [X] Complex code has been commented, including external interfaces
- [ ] Tests have 100% code coverage
- [X] The base branch is either `main`, or there's a description of how to merge

## Issue Resolution

Closes #957.

### CODE REVIEW

File: pkg/pool-hooks/test/foundry/ExitFeeHookExample.t.sol [hunk 1]
```diff
@@ -58,7 +58,8 @@ contract ExitFeeHookExampleTest is BaseVaultTest {
         roleAccounts.poolCreator = lp;
 
         LiquidityManagement memory liquidityManagement;
-        liquidityManagement.disableUnbalancedLiquidity = true;
+        liquidityManagement.disableAddLiquidityUnbalanced = true;
```
Comment (jubeira) [inline pkg/pool-hooks/test/foundry/ExitFeeHookExample.t.sol:61]: This one can actually be enabled right? That's the main use-case for this :)

File: pkg/vault/contracts/VaultExtension.sol [hunk 2]
```diff
@@ -284,7 +287,8 @@ contract VaultExtension is IVaultExtension, VaultCommon, Proxy {
                 // input amount (EXACT_OUT).
                 if (
                     hookFlags.enableHookAdjustedAmounts &&
-                    params.liquidityManagement.disableUnbalancedLiquidity == false
+                    (params.liquidityManagement.disableAddLiquidityUnbalanced == false ||
+                        params.liquidityManagement.disableRemoveLiquidityUnbalanced == false)
```
Comment (jubeira) [inline pkg/vault/contracts/VaultExtension.sol:291]: hm.
The problem is that we want to enable this when the adjusted amount is only for one operation.

E.g. we can enable after exit adjusted amounts with unbalanced adds. Which means that we'd need to also split `hookAdjustedAmounts` flag as well, unless I'm missing something.

**Review by jubeira — COMMENTED** [review]
Thanks for putting this together @EndymionJkb !

Overall looks good for the most part, but I think we'll need to split `hookAdjustedAmounts` flag in two as well to make this work :/

File: pkg/pool-hooks/test/foundry/ExitFeeHookExample.t.sol (hunk 1 above)
Comment (EndymionJkb) [reply pkg/pool-hooks/test/foundry/ExitFeeHookExample.t.sol:61]: Yes, but I didn't want also change the example in this PR (and also update the tests, etc.)

File: pkg/vault/contracts/VaultExtension.sol (hunk 2 above)
Comment (EndymionJkb) [reply pkg/vault/contracts/VaultExtension.sol:291]: We could have hookAdjustedAmountsIn and hookAdjustedAmountsOut. We also check this on swaps, though... do we need three?

File: pkg/pool-weighted/test/gas/.hardhat-snapshots/[WeightedPool - Standard - BatchRouter] swap exact in with one token and fees - cold slots [hunk 3]
```diff
@@ -1 +1 @@
-186.1k
\ No newline at end of file
+186.5k
```
Comment (jubeira) [inline pkg/pool-weighted/test/gas/.hardhat-snapshots/[WeightedPool - Standard - BatchRouter] swap exact in with one token and fees - cold slots:1]: :face_exhaling:

File: pkg/vault/contracts/Vault.sol [hunk 4]
```diff
@@ -646,7 +646,7 @@ contract Vault is IVaultMain, VaultCommon, Proxy {
             bptAmountOut = 0;
             amountsInScaled18 = maxAmountsInScaled18;
         } else if (params.kind == AddLiquidityKind.UNBALANCED) {
```
Comment (jubeira) [inline pkg/vault/contracts/Vault.sol:648]: not strictly related with this PR, but what about donations? Those are also a kind of unbalanced liquidity.

File: pkg/vault/test/.contract-sizes/Vault [hunk 5]
```diff
@@ -1,2 +1,2 @@
-Bytecode	23.883
-InitCode	25.224
\ No newline at end of file
+Bytecode	23.980
```
Comment (jubeira) [inline pkg/vault/test/.contract-sizes/Vault:1]: :melting_face:

File: pkg/vault/test/foundry/unit/VaultExtensionUnit.t.sol [hunk 6]
```diff
@@ -15,6 +15,10 @@ contract VaultExtensionUnitTest is BaseVaultTest {
         BaseVaultTest.setUp();
     }
 
+    function testItWorks() public pure {
+        assertTrue(true);
+    }
```
Comment (jubeira) [inline pkg/vault/test/foundry/unit/VaultExtensionUnit.t.sol:20]: remove?

**Review by jubeira — COMMENTED** [review]
Thanks @EndymionJkb, I think this looks good.

I don't fully buy the extra complexity and gas that this brings tbh. Let's discuss with the rest of the team whether this is worth it or not now that we have a solid implementation and some numbers on the table. Great job!

File: pkg/vault/test/foundry/unit/VaultExtensionUnit.t.sol (hunk 6 above)
Comment (EndymionJkb) [reply pkg/vault/test/foundry/unit/VaultExtensionUnit.t.sol:20]: Ha :) Yes, I do this when debugging an error in the setup.

File: pkg/vault/contracts/Vault.sol (hunk 4 above)
Comment (EndymionJkb) [reply pkg/vault/contracts/Vault.sol:648]: Not sure what you're asking here. We have separate types and flags for donations and regular add unbalanced, and the flags are specific to the operation.

**Comment by EndymionJkb** [comment]
Consensus was this was too complex; close as not planned.

**Pull request closed without merging**